databricks/koalas/mlflow.py

```diff
@@ -0,0 +1,166 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+MLflow-related functions to load models and apply them to Koalas dataframes.
+"""
+from mlflow.pyfunc import PythonModel, load_pyfunc, spark_udf
+from pyspark.sql.types import DataType
+import pandas as pd
+import numpy as np
+
+from databricks.koalas.utils import lazy_property, default_session
+from databricks.koalas import Series, DataFrame
+from databricks.koalas.typedef import as_spark_type
+
+__all__ = ["PythonModelWrapper", "load_model"]
+
+
+class PythonModelWrapper(object):
+    """
+    A wrapper around MLflow's Python object model.
+
+    This wrapper acts as a predictor on koalas
+
+    """
+    def __init__(self, path, run_id, return_type_hint):
+        self._path = path  # type: str
+        self._run_id = run_id  # type: str
+        self._return_type_hint = return_type_hint
+
+    @lazy_property
+    def _return_type(self) -> DataType:
+        hint = self._return_type_hint
+        # The logic is simple for now, because it corresponds to the default
+        # case: continuous predictions
+        # TODO: do something smarter, for example when there is a sklearn.Classifier (it should
+        # return an integer or a categorical)
+        # We can do the same for pytorch/tensorflow/keras models by looking at the output types.
+        # However, this is probably better done in mlflow than here.
+        if hint == 'infer' or not hint:
+            hint = np.float64
+        return as_spark_type(hint)
+
+    @lazy_property
+    def _model(self) -> PythonModel:
+        return load_pyfunc(self._path, self._run_id)
+
+    @lazy_property
+    def _model_udf(self):
+        spark = default_session()
+        return spark_udf(spark, self._path, self._run_id, result_type=self._return_type)
+
+    def __str__(self):
+        return "PythonModelWrapper({})".format(str(self._model))
+
+    def __repr__(self):
+        return "PythonModelWrapper({})".format(repr(self._model))
+
+    def predict(self, data):
+        """
+        Returns a prediction on the data.
+        If the data is a pandas Dataframe, the return is a pandas series (TODO: np.array?)
+        If the data is a koalas DataFrame, the return is a Koalas Series.
+        """
+        if isinstance(data, pd.DataFrame):
+            return self._model.predict(data)
+        if isinstance(data, DataFrame):
+            cols = [data._sdf[n] for n in data.columns]
+            return_col = self._model_udf(*cols)
```

Comment:
Btw, when you apply the udf to *cols, the data is passed to the udf as an array fo columns and the column names are lost which breaks many models. Starting dbr 5.4 conda, you can pass the columns as as a struct in which case the udf gets a dataframe with column names.

Reply:
E.g you can call it like this:
```
spark_df.withColumn("prediction", f(struct("x", "label"))).select("prediction")
```

Or you will be able to after we update our spark_udf call.

Reply:
This is great to know. I just tried and it is not supported yet in OSS Spark <= 2.4.3 (currently the latest). I have put a comment to fix this, along with the code snippet, once spark 2.4.4+ is released.

Reply:
Unfortunately, it won't be included in 2.4 series.

Reply:
Makes sense.  

Unfortunately until that feature becomes available you may be returning different results for pandas dataframe vs spark dataframe. The spark_udf currently returns the left-most numeric column, so if your model predict ran on pandas df returns anything else than 1d numeric vector it's gonna be inconsistent. The return type of mlflow spark_udf  can be controlled by passing in explicit return_type argument. Unfortunately there is no way getting it back as a struct. I would consider copying the same logic here for consistency, but that's not great either.

Reply:
Oh I've you are already passing the return type argument to spark_udf.